### DISCUSSION TITLE
Use `go-vcr` for testing?

### DISCUSSION THREAD

**Discussion by endorama** [elastic/assetbeat#71] [Ideas]
[`dnaeon/go-vcr`](https://github.com/dnaeon/go-vcr) is a library that aids testing API interactions by recording API responses and replaying them in subsequent executions.

As I expect this tool to heavily rely on Cloud Providers API responses at testing, this may be a great addition to lower the complexity of writing tests by reducing the need for manual mocking.

I've been using is in elastic-agent-changelog-tool if you want to observe it's usage ([here](https://github.com/elastic/elastic-agent-changelog-tool/blob/300045fe60c0acd83509fb69d399284283ef7493/internal/githubtest/githubtest.go#L29) the wrapper and [here](https://github.com/elastic/elastic-agent-changelog-tool/blob/300045fe60c0acd83509fb69d399284283ef7493/internal/github/pr-has-labels_test.go#L16-L17) an usage example). The tool interacts with GitHub APIs and go-vcr has been painless to integrate and helpful during development.

**Comment by dmathieu** [comment]
As a comparison/alternative, my GCP PR does it with httptest.
https://github.com/elastic/inputrunner/blob/ff1a554c37c40492163dde2028905145c6c3abc7/input/assets/gcp/compute_test.go#L157-L170

I think one default I'd give to something like go-vcr is that we have data in another file, when we could have everything in the same test file (and only create fixtures for things that are big and would clutter the test file).

**Reply by endorama** [reply]
> I think one default I'd give to something like go-vcr is that we have data in another file, when we could have everything in the same test file (and only create fixtures for things that are big and would clutter the test file).

May you please clarify this? I'm not sure I understood what you mean here.

**Comment by tommyers-elastic** [comment]
i've not come across this tool so thanks for the suggestion! moving to a discussion for now, since nothing directly to action here.
